firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java

```diff
@@ -778,13 +779,43 @@ public void saveNamedQuery(NamedQuery namedQuery, ImmutableSortedSet<DocumentKey
         "Get named query", () -> bundleCache.getNamedQuery(queryName));
   }
 
-  public void configureIndices(List<FieldIndex> fieldIndices) {
+  @VisibleForTesting
+  Collection<FieldIndex> getFieldIndexes() {
+    return persistence.runTransaction("Get indexes", () -> indexManager.getFieldIndexes());
+  }
+
+  public void configureFieldIndexes(List<FieldIndex> newFieldIndexes) {
     persistence.runTransaction(
-        "Configure indices",
+        "Configure indexes",
         () -> {
-          // TODO(indexing): Disable no longer active indices
-          for (FieldIndex fieldIndex : fieldIndices) {
-            indexManager.addFieldIndex(fieldIndex);
+          Collection<FieldIndex> existingIndexes = indexManager.getFieldIndexes();
+
+          // Delete indices that are no longer used.
+          for (FieldIndex existingIndex : existingIndexes) {
```

Comment:
I am a bit worried about performance here. With a 30 indexes setup, this could easily lead up to 1800 (or 1000 realistically) operations. 

I want to suggest to explore something like a bloom filter to help. But I realize this might also be premature optimization. So how about adding a TODO to followup?

Reply:
Also, can the two loop be merged, such that we only call `matchesConstraints` half times?

Reply:
I rewrote the entire thing to run in `O(n log n)`. Can you take another look?